### PR TITLE
Run Kitchen-Docker in TravisCI

### PR DESCRIPTION
**Pull request by jvrplmlmn**
Tests for CentOS will fail until https://github.com/StackStorm/chef-stackstorm/pull/16 gets merged.

Solves #17 

@shortdudey123 @sysbot @armab ?

### CODE REVIEW

**Comment by shortdudey123** [comment]
how about we not remove all the other testing :)

**Comment by jvrplmlmn** [comment]
@shortdudey123 added back `rubocop`, `foodcritic` and `rspec`.

The Kitchen tests are failing for apparently unrelated issues.

**Comment by jvrplmlmn** [comment]
https://stackstorm-community.slack.com/archives/community/p1474226956001831

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
+before_script:
+  # https://github.com/zuazo/kitchen-in-travis-native/issues/1#issuecomment-142455888
+  - sudo iptables -L DOCKER || sudo iptables -N DOCKER
+  - eval "$(/opt/chefdk/bin/chef shell-init bash)"
+  - /opt/chefdk/bin/chef gem install kitchen-docker
+
+script:
+  - /opt/chefdk/bin/chef --version
+  - /opt/chefdk/bin/cookstyle --version
+  - /opt/chefdk/bin/cookstyle
```
Comment (shortdudey123) [inline .travis.yml:26]: Reason for switching? (besides copy and paste 😄)

File: .travis.yml [hunk 2]
```diff
@@ -1,9 +1,31 @@
...
+# Use Travis's cointainer based infrastructure
+sudo: required
+services: docker
+
+addons:
+ apt:
+   sources:
+     - chef-stable-precise
+   packages:
+     - chefdk
```
Comment (shortdudey123) [inline .travis.yml:10]: why do you need `chefdk`? looks like everything is already setup w/o it

File: .travis.yml [hunk 3]
```diff
@@ -1,9 +1,31 @@
...
+
+script:
+  - /opt/chefdk/bin/chef --version
+  - /opt/chefdk/bin/cookstyle --version
+  - /opt/chefdk/bin/cookstyle
+  - /opt/chefdk/bin/foodcritic --version
+  - /opt/chefdk/bin/foodcritic . --exclude spec
+  - /opt/chefdk/embedded/bin/rspec spec
+  - /opt/chefdk/bin/kitchen --version
+  - KITCHEN_LOCAL_YAML=.kitchen.docker.yml /opt/chefdk/bin/kitchen verify ${INSTANCE}
```
Comment (shortdudey123) [inline .travis.yml]: Doesn't destroy at the end, not sure if that matters or not

File: .travis.yml [hunk 4]
```diff
@@ -1,9 +1,31 @@
...
+  - sudo iptables -L DOCKER || sudo iptables -N DOCKER
+  - eval "$(/opt/chefdk/bin/chef shell-init bash)"
+  - /opt/chefdk/bin/chef gem install kitchen-docker
+
+script:
+  - /opt/chefdk/bin/chef --version
+  - /opt/chefdk/bin/cookstyle --version
+  - /opt/chefdk/bin/cookstyle
+  - /opt/chefdk/bin/foodcritic --version
+  - /opt/chefdk/bin/foodcritic . --exclude spec
```
Comment (shortdudey123) [inline .travis.yml:28]: Why change the settings?

**Comment by shortdudey123** [comment]
Since `logrotate` is included in the base install of CentOS and Ubuntu but evidently not the docker images, you should be able to just add it to the `provision_command` section and call it good

**Comment by jvrplmlmn** [comment]
Ran `kitchen` locally and the converge phase fails on the Ubuntu 14.04 box due to issues with the packagecloud repos:

```
(...)
Hit http://archive.ubuntu.com trusty/restricted amd64 Packages
Hit http://archive.ubuntu.com trusty/universe amd64 Packages
Get:2 https://packagecloud.io trusty/main Sources
Fetched 16.8 kB in 10s (1552 B/s)
W: Failed to fetch https://packagecloud.io/StackStorm/stable/ubuntu/dists/trusty/main/source/Sources  Hash Sum mismatch

W: Failed to fetch https://packagecloud.io/StackStorm/stable/ubuntu/dists/trusty/main/binary-amd64/Packages  Hash Sum mismatch

E: Some index files failed to download. They have been ignored, or old ones used instead.
(...)
```

@emedvedev ?

**Comment by shortdudey123** [comment]
@jvrplmlmn see http://blog.packagecloud.io/eng/2016/03/21/apt-hash-sum-mismatch/

File: .travis.yml (hunk 1 above)
Comment (jvrplmlmn) [reply .travis.yml:26]: I'm used to `cookstyle`, which is based on `rubocop`. I think it has saner defaults.

File: .travis.yml (hunk 4 above)
Comment (jvrplmlmn) [reply .travis.yml:28]: Do you mean why we're excluding the `spec` directory?

File: .travis.yml (hunk 2 above)
Comment (jvrplmlmn) [reply .travis.yml:10]: We're installing the `chefdk` to have access to `cookstyle`, `foodcritic`, `rspec` and `kitchen` in the build.

Kitchen itself will install ChefDK in the Docker container where it runs the tests.

File: .travis.yml (hunk 3 above)
Comment (jvrplmlmn) [reply .travis.yml]: This should be indeed `kitchen test` instead of `kitchen verify`. I'll push the change to see how it affects the build in TravisCI.

**Comment by jvrplmlmn** [comment]
@shortdudey123 @sysbot @armab this is up for review. Let me know if you want me to squash the commits.

**Comment by arm4b** [comment]
Thanks @jvrplmlmn for contributing and @shortdudey123 for reviewing, this is nice feature to have indeed!

Green build looks good enough to me :+1: but @shortdudey123 and @sysbot might have better idea about details.

**Review by arm4b — APPROVED** [review]

File: .travis.yml (hunk 1 above)
Comment (shortdudey123) [reply .travis.yml:26]: we should stick with just one of them, not use both

File: .travis.yml (hunk 4 above)
Comment (shortdudey123) [reply .travis.yml:28]: yes

File: recipes/_packages.rb [hunk 5]
```diff
@@ -18,6 +18,22 @@
          'deb'
        end
 
+# Read:
+#  http://blog.packagecloud.io/eng/2016/03/21/apt-hash-sum-mismatch/
+#  https://packagecloud.io/computology/apt-backport
+package 'apt' do
```
Comment (shortdudey123) [inline recipes/_packages.rb:24]: needs spec test

File: recipes/_packages.rb [hunk 6]
```diff
@@ -18,6 +18,22 @@
...
+  version '1.2.10'
+  action :nothing
+end
+
+execute 'Ubuntu Precise and Trusty: APT hash sum mismatch workaround' do
```
Comment (shortdudey123) [inline recipes/_packages.rb:29]: needs spec test

File: .travis.yml (hunk 2 above)
Comment (shortdudey123) [reply .travis.yml:10]: We already installed everything needed except for kitchen

**Review by shortdudey123 — CHANGES_REQUESTED** [review]

File: recipes/_packages.rb [hunk 7]
```diff
@@ -18,6 +18,22 @@
...
+  command 'curl -s https://packagecloud.io/install/repositories/computology/apt-backport/script.deb.sh | bash'
+  only_if { node['platform'] == 'ubuntu' }
+  only_if { %w(trusty precise).include? node['lsb']['codename'] }
```
Comment (shortdudey123) [inline recipes/_packages.rb:32]: combine the 2 `only_if`?

**Comment by arm4b** [comment]
Since the build works in Travis, I think I'll go ahead and merge it to benefit from this PR.
We can change the small things afterwards.

**Comment by arm4b** [comment]
Also, I'd change Docker images to those we already use for testing environment (shipped with `init` systems). See:
https://github.com/StackStorm/ansible-st2/blob/master/.kitchen.yml#L24-L39
https://hub.docker.com/r/stackstorm/packagingtest/tags/

@shortdudey123 Feel free to do after-merge changes as well.

**Comment by arm4b** [comment]
Seeing now how helpful it is, would like to say big Thank You to @jvrplmlmn :tada: for implementation :+1:

**Pull request merged**